### ISSUE TITLE
Trips and Stops detection

### ISSUE THREAD

**Issue by Abyss777** [tananaev/traccar#3248]
Topic to discuss.

You suggested use ignition to detection, I think it can be used only as auxiliary parameter in following way:
If speed is more than threshold but ignition is false interpret this situation as stop.

I believe algorithm will be the same, just `isMoving` calculation be more complex.

May be we can use some other attributes to increase detection precision?

**Comment by tananaev** [comment]
Yes, it would be nice to make it smarter. Agree with using no ignition for detecting stops. Not sure what other parameters we can use.

One thing that we might want to handle is when there is no data for a long time. This is a very common problem. Many devices don't have battery and only send when ignition is ON, which means that Traccar will never detect stop because device basically goes offline straight after turning ignition OFF. I was thinking to make some threshold parameter to treat a long break (e.g. 1 hour by default) in data as a stop.

**Comment by Abyss777** [comment]
"No data" sometime mean no connection for devices that do not have internal cache. But yes we can implement some configurable threshold for detection.

**Comment by tananaev** [comment]
If it's an internal cache, it's OK because we don't save calculate trips anywhere. As soon as all the data is uploaded, we just re-calculate everything.

**Comment by Abyss777** [comment]
I have an idea...

- Introduce obligatory boolean position attribute: `KEY_MOTION` (obligatory like distance and total distance) or even ordinary field.
- Initially it will be set by dedicated pipeline handler according to `speed` and `speedThreshold`
- 'MotionEventHandler` will rise events according to its value (not speed directly)
- Trips and Stops detector will determine motion state also according to this attribute (except detect gaps of course)

The benefit is that it might be corrected by Computed Attributes engine. Some devices has internal motion detector, others has accelerometer connected to ADC. Anyway there might be some additional attributes that help to correct motion state.

Also I remember somebody regretted that we removed `motion` field from device model.
And also it can be used in web client, I remember somebody wanted to colorize device by motion state, not by connection status.

What do you think?

**Comment by tananaev** [comment]
It's a nice idea, but there will be problems like you mentioned in the previous comment. Also, we would need to store more than one field to know when device last moved to calculate stops more accurately, like we do in trips. Also, there will be problem with order or positions like we have with distance.

**Comment by Abyss777** [comment]
It will not help much for Trips and Stops just replace comparing `getSpeed() > speedThreshold` to `getBoolean(Position.KEY_MOTION)`
But its value might be more precise with help of Computed Attributes or even protocol decoder.

Last moved will not help for trips/stop calculation, we need look both in the future and in the past.
Do not think we will get problems like with distance, we are not linked to last position, we just need to know if device is moving or not in the current position.

But not sure now, not so much benefits...

**Comment by tananaev** [comment]
I see what you mean. So, it's not going to be the same as trip stop. Yes, I think then it makes sense.

**Comment by Abyss777** [comment]
`KEY_MOTION` already used https://github.com/tananaev/traccar/blob/master/src/org/traccar/protocol/Gl200ProtocolDecoder.java#L719

But it has other meaning:
>  +RESP:GTNMR:  non movement is detected by motion sensor, according to the setting of AT+GTNMD

I'm going to rename it in GL200 decoder to for example `KEY_NO_MOTION_DETECTED` = `noMotionDetected`
Will be OK?

**Comment by tananaev** [comment]
I think we should keep it. It's exactly what we want it to be (just need to convert to boolean), but returned by the device, so it's going to be much accurate than whatever server can calculate. I think we can use `KEY_MOTION` if device provides it. If not, we calculate it based on the data that we have on the server.

**Comment by Abyss777** [comment]
Sure. I'm going to handle it in this way: provided by device has priority (like distance).
Looks like I just was confused that it is not in every packet and "device config" specific.
